File: slopspotter-virtualization/docker/node/runner.js
const fs = require("fs");
const os = require("os");
const path = require("path");
const { execFileSync, spawnSync } = require("child_process");
const fs = require("fs");

function main() {
  const pkg = process.argv[2];
  if (!pkg) {
    console.log(JSON.stringify({ error: "package name required" }));
    return;
  }

  const work = fs.mkdtempSync(path.join(os.tmpdir(), "scan-"));
const report = {
    package: pkg,
    install_rc: null,
    require_rc: null,
    install_error: null,
    require_error: null,
    installed_version: null,
    download_bytes: null,
    install_out: "",
    install_err: "",
    require_out: "",
    require_err: "",
    network: [],
    processes: [],
    file_ops: [],
    timeout: false,
  };

  try {
    const install = spawnSync(
      "strace",
      [
        "-ff",
        "-e",
        "trace=network,file,process",
        "-o",
        path.join(work, "trace_install"),
        "npm",
        "install",
        pkg,
      ],
      { encoding: "utf8", timeout: 40000 }
    );
    report.install_rc = install.status;
    report.install_out = (install.stdout || "").slice(-4000);
    report.install_err = (install.stderr || "").slice(-4000);
    if (install.status !== 0) {
      report.install_error = "install_failed";
    }
    // Try to read installed version from node_modules/<pkg>/package.json
    try {
      const pkgJsonPath = path.join(process.cwd(), "node_modules", pkg, "package.json");
      if (fs.existsSync(pkgJsonPath)) {
        const pkgJson = JSON.parse(fs.readFileSync(pkgJsonPath, "utf8"));
        if (pkgJson && pkgJson.version) {
          report.installed_version = pkgJson.version;
        }
        // compute dir size
        const dirSize = (dir) => {
          let total = 0;
          const entries = fs.readdirSync(dir, { withFileTypes: true });
          for (const entry of entries) {
            const full = path.join(dir, entry.name);
            if (entry.isDirectory()) total += dirSize(full);
            else total += fs.statSync(full).size;
          }
          return total;
        };
        report.download_bytes = dirSize(path.join(process.cwd(), "node_modules", pkg));
      }
    } catch (e) {
      // ignore
    }

    const req = spawnSync(
      "strace",
      [
        "-ff",
        "-e",
        "trace=network,file,process",
        "-o",
        path.join(work, "trace_require"),
        "node",
        "-e",
        `require("${pkg.replace(/"/g, "")}")`,
      ],
      { encoding: "utf8", timeout: 15000 }
    );
    report.require_rc = req.status;
    report.require_out = (req.stdout || "").slice(-2000);
    report.require_err = (req.stderr || "").slice(-2000);
    if (req.status !== 0) {
      report.require_error = "require_failed";
    }

    fs.readdirSync(work).forEach((fname) => {
      if (!fname.startsWith("trace")) return;
      const file = path.join(work, fname);
      if (!fs.statSync(file).isFile()) return;
      const lines = fs.readFileSync(file, "utf8").split("\n");
      for (const line of lines) {
        if (line.includes("connect(")) report.network.push(line.trim());
        if (line.includes("execve(")) report.processes.push(line.trim());
        if (line.includes("open(") || line.includes("openat(") || line.includes("stat(") || line.includes("access(")) {
          const firstQuote = line.indexOf('"');
          const secondQuote = firstQuote >= 0 ? line.indexOf('"', firstQuote + 1) : -1;
          if (firstQuote >= 0 && secondQuote > firstQuote) {
            const pathVal = line.slice(firstQuote + 1, secondQuote);
            if (pathVal) report.file_ops.push(pathVal);
          }
        }
      }
    });
  } catch (err) {
    if (err.killed) {
      report.timeout = true;
    } else {
      report.install_error = String(err);
    }
  } finally {
    fs.rmSync(work, { recursive: true, force: true });
  }

  console.log(JSON.stringify(report));
}

main();

File: slopspotter-virtualization/docker/python/runner.py
import json
import os
import shutil
import subprocess
import sys
import tempfile
import importlib
import importlib.util


def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "package name required"}))
        return

    pkg = sys.argv[1]
    work = tempfile.mkdtemp(prefix="scan-")
    report = {
        "package": pkg,
        "install_rc": None,
        "import_rc": None,
        "install_error": None,
        "import_error": None,
        "installed_version": None,
        "download_bytes": None,
        "install_out": "",
        "install_err": "",
        "import_out": "",
        "import_err": "",
        "network": [],
        "processes": [],
        "file_ops": [],
        "file_writes": [],
        "timeout": False,
    }

    try:
        install = subprocess.run(
            [
                "strace",
                "-ff",
                "-e",
                "trace=network,file,process",
                "-o",
                os.path.join(work, "trace_install"),
                "pip",
                "install",
                "--no-input",
                "--no-cache-dir",
                pkg,
            ],
            capture_output=True,
            text=True,
            timeout=40,
        )
        report["install_rc"] = install.returncode
        report["install_out"] = (install.stdout or "")[-4000:]
        report["install_err"] = (install.stderr or "")[-4000:]
        if install.returncode != 0:
            report["install_error"] = "install_failed"
        # best-effort parse for version/size from pip output
        for line in (install.stdout or "").splitlines():
            if "Downloading" in line and "(" in line and " kB" in line:
                try:
                    size_part = line.split("(")[-1].split(" ")[0]
                    report["download_bytes"] = int(float(size_part) * 1024)
                except Exception:
                    pass
            if "Installing collected packages" in line:
                parts = line.split()
                if len(parts) >= 3:
                    report["installed_version"] = parts[-1]
        # Try to read installed version and approximate size from site-packages
        try:
            import importlib.metadata as metadata

            if not report["installed_version"]:
                report["installed_version"] = metadata.version(pkg)
            dist = metadata.distribution(pkg)
            if dist and dist.locate_file(""):
                base_path = dist.locate_file("").parent
                total = 0
                for root, _, files in os.walk(base_path):
                    for f in files:
                        fp = os.path.join(root, f)
                        try:
                            total += os.path.getsize(fp)
                        except OSError:
                            continue
                if total:
                    report["download_bytes"] = total
        except Exception:
            pass

        mod_name = pkg.replace("-", "_").split(".")[0]
        imp = subprocess.run(
            [
                "strace",
                "-ff",
                "-e",
                "trace=network,file,process",
                "-o",
                os.path.join(work, "trace_import"),
                "python",
                "-c",
                f"import {mod_name}",
            ],
            capture_output=True,
            text=True,
            timeout=15,
        )
        report["import_rc"] = imp.returncode
        report["import_out"] = (imp.stdout or "")[-2000:]
        report["import_err"] = (imp.stderr or "")[-2000:]
        if imp.returncode != 0:
            report["import_error"] = "import_failed"

        for fname in os.listdir(work):
            if not fname.startswith("trace"):
                continue
            path = os.path.join(work, fname)
            if not os.path.isfile(path):
                continue
            with open(path, "r", errors="ignore") as fh:
                for line in fh:
                    if "connect(" in line:
                        report["network"].append(line.strip())
                    if "execve(" in line:
                        report["processes"].append(line.strip())
                    if any(tok in line for tok in ("open(", "openat(", "stat(", "access(")):
                        # Capture file paths inside quotes, if present
                        parts = line.split("\"")
                        if len(parts) >= 2:
                            path_val = parts[1]
                            if path_val:
                                report["file_ops"].append(path_val)
                                # Heuristic: consider write if flags include O_WRONLY/O_RDWR/O_CREAT/O_TRUNC/O_APPEND
                                if any(flag in line for flag in ("O_WRONLY", "O_RDWR", "O_CREAT", "O_TRUNC", "O_APPEND")):
                                    report["file_writes"].append(path_val)
    except subprocess.TimeoutExpired:
        report["timeout"] = True
    finally:
        shutil.rmtree(work, ignore_errors=True)

    print(json.dumps(report))


if __name__ == "__main__":
    main()
